pkg/storegateway/bucket_test.go

```diff
@@ -1929,6 +1930,125 @@ func TestSeries_BlockWithMultipleChunks(t *testing.T) {
+	require.NoError(t, err)
+
+	_, err = testhelper.CreateBlock(ctx, bktDir, []labels.Labels{
+		labels.FromStrings(labels.MetricName, "series_1"),
+		labels.FromStrings(labels.MetricName, "series_2"),
+	}, numSamplesPerSeries, minTime, maxTime, nil, 0)
+	require.NoError(t, err)
+
+	// Create a bucket and upload the block there.
+	bkt, err := filesystem.NewBucket(bktDir)
+	assert.NoError(t, err)
+	defer func() { assert.NoError(t, bkt.Close()) }()
+
+	instrBkt := objstore.WithNoopInstr(bkt)
+
+	// Instance a real bucket store we'll use to query the series.
+	fetcher, err := block.NewMetaFetcher(logger, 10, instrBkt, tmpDir, nil, nil)
+	assert.NoError(t, err)
+
+	tests := map[string]struct {
+		reqMatchers    []storepb.LabelMatcher
+		seriesLimit    uint64
+		chunksLimit    uint64
+		expectedErr    string
+		expectedSeries int
+	}{
+		"should fail if the number of unique series queried is less than the configured series limit": {
+			reqMatchers: []storepb.LabelMatcher{{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_[12]"}},
+			seriesLimit: 1,
+			expectedErr: errSeriesLimitMessage,
+		},
+		"should pass if the number of unique series queried is equal or greater than the configured series limit": {
+			reqMatchers:    []storepb.LabelMatcher{{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_[12]"}},
+			seriesLimit:    2,
+			expectedSeries: 2,
+		},
+		"should fail if the number of chunks queried is less than the configured chunks limit": {
+			reqMatchers: []storepb.LabelMatcher{{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_[12]"}},
+			chunksLimit: 3,
+			expectedErr: errChunksLimitMessage,
+		},
+		"should pass if the number of chunks queried is equal or greater than the configured chunks limit": {
+			reqMatchers:    []storepb.LabelMatcher{{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_[12]"}},
+			chunksLimit:    4,
+			expectedSeries: 2,
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			for _, batchSize := range []int{1, 2} {
+				t.Run(fmt.Sprintf("batch size: %d", batchSize), func(t *testing.T) {
+					store, err := NewBucketStore(
+						"tenant",
+						instrBkt,
+						fetcher,
+						tmpDir,
+						NewChunksLimiterFactory(testData.seriesLimit),
+						NewSeriesLimiterFactory(testData.chunksLimit),
+						newGapBasedPartitioner(mimir_tsdb.DefaultPartitionerMaxGapSize, nil),
+						10,
+						mimir_tsdb.DefaultPostingOffsetInMemorySampling,
+						indexheader.Config{},
+						false,
+						0,
+						hashcache.NewSeriesHashCache(1024*1024),
+						NewBucketStoreMetrics(nil),
+						WithStreamingSeriesPerBatch(batchSize),
+					)
+					assert.NoError(t, err)
+					assert.NoError(t, store.SyncBlocks(ctx))
+
+					// Override the limiters based on the configured limits in the test case.
+					store.seriesLimiterFactory = NewSeriesLimiterFactory(testData.seriesLimit)
```

Comment:
Are lines 2027 and 2028 really needed? Don't lines 2011 and 2012 already do the job?

Reply:
You're right. It comes from a refactoring I did after writing this test. Will fix it.